### PR TITLE
use split instead of replace for windows case

### PR DESCRIPTION
**Pull request by KazuCocoa**
closes https://github.com/appium/appium-doctor/issues/33

On Windows environment, we can get multiple lines as the result of `where node`.
The 1st line is expected node. On *NIX environment like mac, `which` returns one line. This logic can work both environment.

```
$ where node

F:\Tools\nodejs\node.exe
C:\Program Files\nodejs\node.exe
C:\Program Files\Microsoft MPI\Bin\node.exe
```

### CODE REVIEW

File: lib/node-detector.js [hunk 1]
```diff
@@ -71,7 +72,7 @@ class NodeDetector {
       log.debug(err);
       return null;
     }
-    let nodePath = stdout.replace("\n", "");
+    let nodePath = stdout.split(/\r?\n/)[0];
```
Comment (KazuCocoa) [inline lib/node-detector.js]: `''.split(/\r?\n/)[0]; //=> ''` and it fails in next line

File: lib/node-detector.js [hunk 2]
```diff
@@ -36,7 +36,8 @@ class NodeDetector {
       log.debug(err);
       return null;
     }
-    let nodePath = stdout.replace(/[\n\r]/g, "");
+    // stdout can be multiple lines in windows environment. Read https://github.com/appium/appium-doctor/issues/33
+    let nodePath = stdout.split(/\r?\n/)[0];
```
Comment (mykola-mokhnach) [inline lib/node-detector.js]: There is os.EOL constant for this purpose

Comment (mykola-mokhnach) [reply lib/node-detector.js]: this will throw an exception if stdout is empty

File: lib/node-detector.js (hunk 1 above)
Comment (mykola-mokhnach) [inline lib/node-detector.js]: it probably makes sense to extract path extraction into a separate helper function

File: lib/node-detector.js (hunk 2 above)
Comment (KazuCocoa) [reply lib/node-detector.js]: https://nodejs.org/api/os.html#os_os_eol oh, it's helpful

File: lib/node-detector.js (hunk 1 above)
Comment (KazuCocoa) [reply lib/node-detector.js]: This line was already filtered as macOS
https://github.com/appium/appium-doctor/pull/45/files#diff-5cee4c29bc65633956405e495b2a9f4aR44

File: lib/dev.js [hunk 3]
```diff
@@ -15,17 +14,11 @@ class BinaryIsInPathCheck extends DoctorCheck {
...
-    try {
-      let executable = system.isWindows() ? 'where' : 'which';
-      let {stdout} = await exec(executable, [this.binary]);
-      if (stdout.match(/not found/gi)) {
-        throw new Error('Not Found');
-      }
-      resolvedPath = system.isWindows() ? stdout.split(EOL)[0] : stdout.replace(EOL, '');
-    } catch (err) {
+    const resolvedPath = await resolveExecutablePath(this.binary);
+    if (_.isNull(resolvedPath)) {
```
Comment (mykola-mokhnach) [inline lib/dev.js]: I would simply test `if (resolvedPath)`

File: lib/node-detector.js [hunk 4]
```diff
@@ -23,25 +25,17 @@ class NodeDetector {
...
+    if (_.isNull(nodePath)) {
       return null;
     }
-    let nodePath = stdout.replace(/[\n\r]/g, "");
+
     if (await fs.exists(nodePath)) {
-      log.debug(`Node binary found using ${cmd} command at: ${nodePath}`);
+      log.debug(`Node binary found at: ${nodePath}`);
       return nodePath;
     } else {
```
Comment (mykola-mokhnach) [inline lib/node-detector.js]: else is redundant

File: lib/utils.js [hunk 5]
```diff
@@ -30,4 +33,31 @@ function configureBinaryLog (opts) {
...
+ * @param {string} cmd Standard output by command
+ * @return {?string} The full path of cmd. `null` if the cmd is not found.
+ */
+async function resolveExecutablePath (cmd) {
+  const executable = system.isWindows() ? 'where' : 'which';
+  let stdout;
+  try {
+    stdout = (await exec(executable, [cmd])).stdout;
+    if (stdout.match(/not found/gi)) {
+      throw new Error('Not Found');
```
Comment (mykola-mokhnach) [inline lib/utils.js]: what is the point to throw an error and then catch it?

File: lib/utils.js [hunk 6]
```diff
@@ -30,4 +33,31 @@ function configureBinaryLog (opts) {
...
+    }
+  } catch (err) {
+    log.debug(err);
+    return null;
+  }
+
+  if (_.isString(stdout)) {
+    return system.isWindows() ? stdout.split(EOL)[0] : stdout.replace(EOL, '');
+  } else {
```
Comment (mykola-mokhnach) [inline lib/utils.js]: else is redundant

File: lib/utils.js [hunk 7]
```diff
@@ -30,4 +33,31 @@ function configureBinaryLog (opts) {
   log.level = opts.debug ? 'debug' : 'info';
 }
 
-export { pkgRoot, ok, nok, inquirer, configureBinaryLog, authorizeIos, };
+/**
+ * Return an executable path of cmd
+ *
+ * @param {string} cmd Standard output by command
+ * @return {?string} The full path of cmd. `null` if the cmd is not found.
```
Comment (mykola-mokhnach) [inline lib/utils.js]: throws description is missing

File: lib/utils.js (hunk 5 above)
Comment (KazuCocoa) [reply lib/utils.js]: I figure out windows returns `not found` message when the command cannot find it.
But the message is localised one. It works only for English environment.

I'll change like below

```
if (!stdout.includes(cmd)) {
    return null;
}
```

File: lib/utils.js [hunk 8]
```diff
@@ -30,4 +33,32 @@ function configureBinaryLog (opts) {
...
+    const stdout = (await exec(executable, [cmd])).stdout;
+    if (_.isString(stdout)) {
+      const executablePath = system.isWindows() ? stdout.split(EOL)[0] : stdout.replace(EOL, '');
+      if (await fs.exists(executablePath)) {
+        return executablePath;
+      } else {
+        log.debug(`No executable path of '${cmd}'`);
+        return null;
+      }
+    } else {
```
Comment (mykola-mokhnach) [inline lib/utils.js]: else is redundant. `return null` can be put at the end of the function

File: lib/utils.js [hunk 9]
```diff
@@ -41,24 +41,20 @@ function configureBinaryLog (opts) {
  */
 async function resolveExecutablePath (cmd) {
   const executable = system.isWindows() ? 'where' : 'which';
+  let stdout;
   try {
-    const stdout = (await exec(executable, [cmd])).stdout;
+    stdout = (await exec(executable, [cmd])).stdout;
```
Comment (mykola-mokhnach) [inline lib/utils.js]: this could be also changed to `({stdout} = await exec(executable, [cmd])`

File: lib/utils.js [hunk 10]
```diff
@@ -41,24 +41,20 @@ function configureBinaryLog (opts) {
...
       }
-    } else {
-      log.debug(`'${stdout}' is not string`);
-      return null;
     }
   } catch (err) {
     log.debug(err);
-    return null;
   }
+  log.debug(`No executable path of '${cmd}'. The output of where/which is '${stdout}'`);
```
Comment (mykola-mokhnach) [inline lib/utils.js]: `The output of ${executable}`

File: lib/utils.js [hunk 11]
```diff
@@ -41,24 +41,20 @@ function configureBinaryLog (opts) {
  */
 async function resolveExecutablePath (cmd) {
   const executable = system.isWindows() ? 'where' : 'which';
+  let stdout;
   try {
-    const stdout = (await exec(executable, [cmd])).stdout;
+    stdout = (await exec(executable, [cmd])).stdout;
     if (_.isString(stdout)) {
```
Comment (mykola-mokhnach) [inline lib/utils.js]: if (stdout)

File: lib/utils.js [hunk 12]
```diff
@@ -41,24 +41,20 @@ function configureBinaryLog (opts) {
...
-      } else {
-        log.debug(`No executable path of '${cmd}'`);
-        return null;
       }
-    } else {
-      log.debug(`'${stdout}' is not string`);
-      return null;
     }
   } catch (err) {
     log.debug(err);
```
Comment (mykola-mokhnach) [inline lib/utils.js]: warn

File: lib/dev.js [hunk 13]
```diff
@@ -15,19 +13,12 @@ class BinaryIsInPathCheck extends DoctorCheck {
...
-      let executable = system.isWindows() ? 'where' : 'which';
-      let {stdout} = await exec(executable, [this.binary]);
-      if (stdout.match(/not found/gi)) {
-        throw new Error('Not Found');
-      }
-      resolvedPath = system.isWindows() ? stdout.split(EOL)[0] : stdout.replace(EOL, '');
-    } catch (err) {
+    const resolvedPath = await resolveExecutablePath(this.binary);
+    if (!resolvedPath) {
       return nok(`${this.binary} is MISSING in PATH!`);
```
Comment (mykola-mokhnach) [inline lib/dev.js]: I think it might be useful for clients to show the actual PATH value in the error message

File: lib/utils.js [hunk 14]
```diff
@@ -30,4 +33,28 @@ function configureBinaryLog (opts) {
...
+ * @param {string} cmd Standard output by command
+ * @return {?string} The full path of cmd. `null` if the cmd is not found.
+ */
+async function resolveExecutablePath (cmd) {
+  const executable = system.isWindows() ? 'where' : 'which';
+  let stdout;
+  try {
+    stdout = (await exec(executable, [cmd])).stdout;
+    if (_.isString(stdout)) {
+      const executablePath = system.isWindows() ? stdout.split(EOL)[0] : stdout.replace(EOL, '');
```
Comment (mykola-mokhnach) [inline lib/utils.js]: doesn't `stdout.replace(EOL, '')` performs the same as `trim()`?

**Pull request merged**